src/provider/provider_tracking.c

```diff
@@ -216,6 +260,116 @@ static const char *trackingName(void *provider) {
+        (umf_tracking_memory_provider_t *)provider;
+    return umfMemoryProviderGetIPCHandleSize(p->hUpstream, size);
+}
+
+static enum umf_result_t trackingGetIpcHandle(void *provider, const void *ptr,
+                                              size_t size, void *ipcData) {
+    umf_tracking_memory_provider_t *p =
+        (umf_tracking_memory_provider_t *)provider;
+    enum umf_result_t ret = UMF_RESULT_SUCCESS;
+    size_t ipcDataSize = 0;
+    int cached = 0;
+    do {
+        void *value = critnib_get(p->ipcCache, (uintptr_t)ptr);
+        if (value) { //cache hit
+            struct ipc_cache_value_t *cache_value =
+                (struct ipc_cache_value_t *)value;
+            memcpy(ipcData, cache_value->ipcData, cache_value->size);
+            cached = 1;
+        } else {
+            ret =
+                umfMemoryProviderGetIPCHandle(p->hUpstream, ptr, size, ipcData);
+            if (ret != UMF_RESULT_SUCCESS) {
+                return ret;
+            }
+
+            ret = umfMemoryProviderGetIPCHandleSize(p->hUpstream, &ipcDataSize);
+            assert(ret == UMF_RESULT_SUCCESS);
+
+            size_t value_size = sizeof(struct ipc_cache_value_t) + ipcDataSize;
+            struct ipc_cache_value_t *cache_value =
+                (struct ipc_cache_value_t *)malloc(value_size);
+
+            cache_value->size = ipcDataSize;
+            memcpy(cache_value->ipcData, ipcData, ipcDataSize);
+
+            int insRes = critnib_insert(p->ipcCache, (uintptr_t)ptr,
+                                        (void *)cache_value, 0 /*update*/);
+            if (insRes == 0) {
+                cached = 1;
+            } else {
+                // critnib_insert might fail in 2 cases:
+                // 1. Another thread created cache entry. So we need to
+                //    clean up allocated handle and try to read again from
+                //    the cache. Alternative approach could be insert empty
+                //    cache_value and only if insert succeed get actual IPC
+                //    handle and fill the cache_value structure under the lock.
+                //    But this case should be rare enough.
+                // 2. critnib failed to allocate memory internally. We need
+                //    to cleanup and return corresponding error.
+                free(cache_value);
+                ret = umfMemoryProviderPutIPCHandle(p->hUpstream, ipcData);
+                if (ret != UMF_RESULT_SUCCESS) {
+                    return ret;
+                }
+                if (insRes == ENOMEM) {
+                    return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+                }
+            }
+        }
+    } while (!cached);
+
+    return ret;
+}
+
+static enum umf_result_t trackingPutIpcHandle(void *provider, void *ipcData) {
+    (void)provider;
+    (void)ipcData;
+    // We just keep ipcData in the provider->ipcCache.
+    // The actual Put is called inside trackingFree
+    return UMF_RESULT_SUCCESS;
```

Comment:
If we never remove from the cache, perhaps it would make sense to have some limit on the size of the cache?

Reply:
Yeah, I considered it. We need real cache implementation with LRU or LFU policy instead of just map. We can do it later if we will see that real use cases require this